### PR TITLE
[SMD-386]: make cols nullable depending on if project is complete

### PR DESCRIPTION
**Pull request by Joel-Pearce**
When a project is complete, "Currently Project Delivery Stage" and columns pertaining to upcoming comms milestones are now nullable.

- [ ] Unit tests and other appropriate tests added or updated
- [ ] README and other documentation has been updated / added (if needed)
- [ ] Commit messages are meaningful and follow good commit message guidelines (e.g. "FS-XXXX: Add margin to nav items preventing overlapping of logo")

### CODE REVIEW

File: core/validation/specific_validations/towns_fund_round_four.py [hunk 1]
```diff
@@ -500,6 +501,70 @@ def validate_postcodes(workbook: dict[str, pd.DataFrame]) -> list["TownsFundRoun
...
+
+    :param df: DataFrame to search for invalid rows.
+    :param column: Column name to check for null values.
+    :param na_values: Set of values considered as null.
+    :return: DataFrame containing rows with null values in the specified column.
+    """
+    invalid_rows = df[df[column].isin(na_values)]
+    if len(invalid_rows) > 0:
+        invalid_rows["null_column"] = column
+    return invalid_rows
```
Comment (Connor-OS) [inline core/validation/specific_validations/towns_fund_round_four.py:565]: I would be tempted to re-factor this so to avoid altering the dataframe rows.
```
invalid_rows =[]
for row in invalid_rows:
	for column in columns_to_check:
		if row[column].isin(na_values):
			invalid_rows.append(row.name, column)

return invalid_rows
```

**Review by Connor-OS — CHANGES_REQUESTED** [review]
Looks good but I'm not sure it always picks up all blank cells that are pd.NA

File: core/validation/specific_validations/towns_fund_round_four.py (hunk 1 above)
Comment (Connor-OS) [reply core/validation/specific_validations/towns_fund_round_four.py:565]: potentially less readably though

File: core/validation/specific_validations/towns_fund_round_four.py [hunk 2]
```diff
@@ -500,6 +501,70 @@ def validate_postcodes(workbook: dict[str, pd.DataFrame]) -> list["TownsFundRoun
...
 
 
+def validate_current_project_delivery_status_and_upcoming_comms_milestones(
+    workbook: dict[str, pd.DataFrame]
+) -> list["TownsFundRoundFourValidationFailure"] | None:
+    """
+    Validates the Project Progress table in the Round 4 submission workbook.
+
+    Checks for rows with a Project Delivery Status other than "4. Completed" where both
+    the Current Project Delivery Stage and Most Important Upcoming Comms Milestone are null.
```
[the remark this answers is not in the document]
Comment (Joel-Pearce) [reply core/validation/specific_validations/towns_fund_round_four.py:511]: Yes, that's probably more accurate.

File: core/validation/specific_validations/towns_fund_round_four.py [hunk 3]
```diff
@@ -500,6 +501,70 @@ def validate_postcodes(workbook: dict[str, pd.DataFrame]) -> list["TownsFundRoun
...
+    :param workbook: A dictionary containing sheet names as keys and pandas DataFrames
+                     representing each sheet in the Round 4 submission.
+    :return: A list of TownsFundRoundFourValidationFailure instances if validation failures occur,
+             otherwise returns None.
+    """
+    project_progress_df = workbook["Project Progress"]
+
+    project_incomplete_rows = project_progress_df[project_progress_df["Project Delivery Status"] != "4. Completed"]
+
+    na_values = {"", "< Select >", np.nan, None}
```
[the remark this answers is not in the document]
Comment (Joel-Pearce) [reply core/validation/specific_validations/towns_fund_round_four.py:523]: I think it can, I suppose the question is whether we want to have the ```na_values``` adjustable or not. For example, for dropdowns we want "< Select >" to be null, for non-dropdowns we don't both with the value. There might be a more useful use case that arises in future development, not sure.

Are there instances where np.nan will not catch a null value that is only picked up by pd.NA? In our extraction scripts we default null values to np.nan. Have you discovered null values appearing as pd.NA? If so we need to change the code for all null validation.

File: core/validation/specific_validations/towns_fund_round_four.py (hunk 1 above)
Comment (Joel-Pearce) [reply core/validation/specific_validations/towns_fund_round_four.py:565]: We need the tacked-on ```"null_column"``` for our specific use case in which there are three columns we're assessing, but one or more of them could be wrong on the same line. As each validation failure needs to return the specific column where the error has occurred, we need a way of distinguishing from which column the failure has occurred if there are errors in multiple columns in one row. This is a workaround as the alternative seemed to be three separate DataFrames over which we iterate three times.

With that in mind, I don't think the above code woks?

File: core/validation/specific_validations/towns_fund_round_four.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (cyrusdobbs) [reply core/validation/specific_validations/towns_fund_round_four.py:523]: Correct me if I'm wrong but `< Select >` is removed at transformation time I think?

File: core/validation/specific_validations/towns_fund_round_four.py [hunk 4]
```diff
@@ -500,6 +501,70 @@ def validate_postcodes(workbook: dict[str, pd.DataFrame]) -> list["TownsFundRoun
...
+    columns_to_check = [
+        "Current Project Delivery Stage",
+        "Most Important Upcoming Comms Milestone",
+        "Date of Most Important Upcoming Comms Milestone (e.g. Dec-22)",
+    ]
+
+    invalid_rows = pd.DataFrame()
+
+    for column in columns_to_check:
+        invalid_rows = pd.concat([invalid_rows, find_invalid_rows(project_incomplete_rows, column, na_values)], axis=0)
```
Comment (cyrusdobbs) [inline core/validation/specific_validations/towns_fund_round_four.py:534]: This is quite hard to follow and is a new way of doing this check. Can it be simplified?

File: core/validation_schema.py [hunk 5]
```diff
@@ -144,17 +144,16 @@
...
             "Project Adjustment Request Status",
             "Delivery (RAG)",
             "Spend (RAG)",
             "Risk (RAG)",
             "Commentary on Status and RAG Ratings",
-            "Most Important Upcoming Comms Milestone",
-            "Date of Most Important Upcoming Comms Milestone (e.g. Dec-22)",
-            # TODO: Added all fields for Round 4 (was previously only Project ID)
+            # Most Important Upcoming Comms Milestone is sometimes nullable so has its own specific validation
+            # This also applies to Date of Most Important Upcoming Comms Milestone (e.g. Dec-22)
```
Comment (cyrusdobbs) [inline core/validation_schema.py:156]: I think this is another use case for splitting dataframes up in a previous stage. We could have `ProjectProgress` and `ProjectProgressComplete` - and then we could use the schema validation as usual? Not one for now just a thought

File: core/validation/specific_validations/towns_fund_round_four.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (cyrusdobbs) [reply core/validation/specific_validations/towns_fund_round_four.py:523]: I agree with Connor here - I think we want consistency. Do we create a set of NA values like this anywhere else? Can we define once and reference instead if so

**Review by cyrusdobbs — CHANGES_REQUESTED** [review]

File: core/validation/messages.py [hunk 6]
```diff
@@ -2,6 +2,7 @@
 
 # BLANK MESSAGES
 BLANK = "The cell is blank but is required."
+BLANK_IF_PROJECT_INCOMPLETE = "The cell is blank but is required for incomplete projects."
```
[the remark this answers is not in the document]
Comment (Joel-Pearce) [reply core/validation/messages.py:5]: It has come from design.

File: core/validation_schema.py (hunk 5 above)
Comment (Joel-Pearce) [reply core/validation_schema.py:156]: I'm not sure how straightforward that would be. Splitting DataFrames makes a lot of sense when we're retaining the tables as they original appear in the spreadsheet (like footfall and non-footfall outcomes), but splitting up a 'natural' DataFrame into two artificial ones based on how our validation works might be less clear than the bespoke validation. One to look at I suppose

Comment (cyrusdobbs) [reply core/validation_schema.py:156]: Hmmm maybe you're right. We need to try to move away from this kind of specific validation though and have a repeatable, configurable way of doing it.. Moving to a more simple spreadsheet built by us will help but I think this kind of scenario could arise then as well potentially.

File: core/validation/specific_validations/towns_fund_round_four.py [hunk 7]
```diff
@@ -500,6 +500,70 @@ def validate_postcodes(workbook: dict[str, pd.DataFrame]) -> list["TownsFundRoun
...
+            TownsFundRoundFourValidationFailure(
+                sheet="Project Progress",
+                section="Projects Progress Summary",
+                column=row["null_column"],
+                message=msgs.BLANK_IF_PROJECT_INCOMPLETE,
+                row_indexes=[row.name],
+            )
+            for _, row in invalid_rows.iterrows()
+        ]
+
```
Comment (cyrusdobbs) [inline core/validation/specific_validations/towns_fund_round_four.py:547]: I think creating a new df and doing this concatenation is hard to follow and doesn't offer any benefit.


I propose this way instead (with `find_null_values` in utils):
```    
def validate_current_project_delivery_status_and_upcoming_comms_milestones(
    workbook: dict[str, pd.DataFrame]
) -> list["TownsFundRoundFourValidationFailure"]:
    """
    Validates the Project Progress table in the Round 4 submission workbook.

    Checks for rows with a Project Delivery Status other than "4. Completed" where both
    the Current Project Delivery Stage or Most Important Upcoming Comms Milestone are null.
    If such invalid states are found, returns a list of TownsFundRoundFourValidationFailure instances.

    :param workbook: A dictionary containing sheet names as keys and pandas DataFrames
                     representing each sheet in the Round 4 submission.
    :return: A list of TownsFundRoundFourValidationFailure instances if validation failures occur,
             otherwise returns None.
    """
    project_progress_df = workbook["Project Progress"]

    project_incomplete_rows = project_progress_df[project_progress_df["Project Delivery Status"] != "4. Completed"]

    columns_to_check = [
        "Current Project Delivery Stage",
        "Most Important Upcoming Comms Milestone",
        "Date of Most Important Upcoming Comms Milestone (e.g. Dec-22)",
    ]

    failures = []
    for column in columns_to_check:
        invalid_rows = find_null_values(project_incomplete_rows, column)
        for _, row in invalid_rows.iterrows():
            failures.append(
                TownsFundRoundFourValidationFailure(
                    sheet="Project Progress",
                    section="Projects Progress Summary",
                    column=column,
                    message=msgs.BLANK_IF_PROJECT_INCOMPLETE,
                    row_indexes=[row.name],
                )
            )

    return failures


def find_null_values(df: pd.DataFrame, column: str) -> pd.DataFrame:
    """Helper function to find rows with null values in a specified column.

    :param df: DataFrame to search for invalid rows.
    :param column: Column name to check for null values.
    :return: DataFrame of offending rows
    """
    return df[df[column].isin(NA_VALUES)]
```

**Review by cyrusdobbs — CHANGES_REQUESTED** [review]

File: core/validation/specific_validations/towns_fund_round_four.py [hunk 8]
```diff
@@ -500,6 +500,70 @@ def validate_postcodes(workbook: dict[str, pd.DataFrame]) -> list["TownsFundRoun
...
+    Checks for rows with a Project Delivery Status other than "4. Completed" where both
+    the Current Project Delivery Stage or Most Important Upcoming Comms Milestone are null.
+    If such invalid states are found, returns a list of TownsFundRoundFourValidationFailure instances.
+
+    :param workbook: A dictionary containing sheet names as keys and pandas DataFrames
+                     representing each sheet in the Round 4 submission.
+    :return: A list of TownsFundRoundFourValidationFailure instances if validation failures occur,
+             otherwise returns None.
+    """
+    project_progress_df = workbook["Project Progress"]
```
[the remark this answers is not in the document]
Comment (Joel-Pearce) [reply core/validation/specific_validations/towns_fund_round_four.py:518]: I think it can, there's probably a way to refactor ```delayed_rows``` into ```columns_to_check```, I'll look at it now

File: core/validation/specific_validations/towns_fund_round_four.py (hunk 7 above)
Comment (Joel-Pearce) [reply core/validation/specific_validations/towns_fund_round_four.py:547]: Makes sense, especially if can refactor leading_factor_of_delay into the cols to check

Comment (Joel-Pearce) [reply core/validation/specific_validations/towns_fund_round_four.py]: The two functions require slightly different rows to check in and slightly different error messages, but can work if as below:

```
# the column to check alongside the rows it should be checked in and the failure message
    columns_to_check = [
        ("Leading Factor of Delay", project_delayed_rows, msgs.BLANK),
        ("Current Project Delivery Stage", project_incomplete_rows, msgs.BLANK_IF_PROJECT_INCOMPLETE),
        ("Most Important Upcoming Comms Milestone", project_incomplete_rows, msgs.BLANK_IF_PROJECT_INCOMPLETE),
        ("Date of Most Important Upcoming Comms Milestone (e.g. Dec-22)", project_incomplete_rows,
         msgs.BLANK_IF_PROJECT_INCOMPLETE),
    ]

    failures = []
    for column, invalid_rows, message in columns_to_check:
        invalid_rows = find_null_values(invalid_rows, column)
        for _, row in invalid_rows.iterrows():
            failures.append(
                TownsFundRoundFourValidationFailure(
                    sheet="Project Progress",
                    section="Projects Progress Summary",
                    column=column,
                    message=message,
                    row_indexes=[row.name],
                )
            )

    return failures
    ```

File: core/validation/const.py [hunk 9]
```diff
@@ -0,0 +1,5 @@
+import numpy as np
+import pandas as pd
+
+# values we treat as null
+NA_VALUES = {"", "< Select >", np.nan, None, pd.NA}
```
Comment (cyrusdobbs) [inline core/validation/const.py:5]: Do we actually want `< Select >` here?

File: core/validation/utils.py [hunk 10]
```diff
@@ -12,3 +14,21 @@ def remove_duplicate_indexes(df: pd.DataFrame):
...
+    if this function is called for each column to be validated.
+
+    :param df: DataFrame to search for invalid rows.
+    :param column: Column name to check for null values.
+    :return: DataFrame containing rows with null values in the specified column.
+    """
+    invalid_rows = df[df[column].isin(NA_VALUES)]
+    if len(invalid_rows) > 0:
+        invalid_rows["null_column"] = column
+    return invalid_rows
```
Comment (cyrusdobbs) [inline core/validation/utils.py:34]: We don't need to set a new column. Can just be this:

```
def find_null_values(df: pd.DataFrame, column: str) -> pd.DataFrame:
    """Helper function to find rows with null values in a specified column.

    :param df: DataFrame to search for invalid rows.
    :param column: Column name to check for null values.
    :return: DataFrame of offending rows
    """
    return df[df[column].isin(NA_VALUES)]
```

**Review by cyrusdobbs — COMMENTED** [review]
Looking good! Just 2 minor comments

File: core/validation/specific_validations/towns_fund_round_four.py (hunk 7 above)
Comment (cyrusdobbs) [reply core/validation/specific_validations/towns_fund_round_four.py]: Looks good

File: core/validation/const.py (hunk 9 above)
Comment (Joel-Pearce) [reply core/validation/const.py:5]: It was written into the tests for leading_factor_of_delay. Looking at the extraction function for project progress, we replace all instances with an empty string. I've now removed all mention of < Select > from the validation

File: tests/test_misc.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    invalid_rows = find_null_values(df, column)
+
+    expected_invalid_rows = pd.DataFrame(
+        index=[7],
+        data=[
+            {"not-nullable": "", "nullable": ""},
+        ],
+    )
+
+    assert_frame_equal(invalid_rows, expected_invalid_rows)
```
Comment (cyrusdobbs) [inline tests/test_misc.py:49]: These are best placed in `tests/validation_tests/test_utils` instead I think

Comment (Joel-Pearce) [reply tests/test_misc.py]: Makes sense, I've moved the tests and pushed

**Review by cyrusdobbs — APPROVED** [review]
LGTM

**Review by Connor-OS — APPROVED** [review]
Looks good
